Add signal-to-noise threshold to PDCalibration

bolotovskyr opened the pull request:
**Description of work.**

The issue: https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=245

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

<!-- Instructions for testing. -->

<!--Fixes #xxxx.--> <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

From mantid workbench, run a modified version of the script originally provided by [66j](https://code.ornl.gov/66j):
```
from mantid.simpleapi import *
LoadNexus(Filename='/SNS/SNAP/shared/Malcolm/Temp/calFile_8x8.nxs',OutputWorkspace='calFile_8x8')
peaksToFit = '2.0346,1.7620,1.2459,1.0625,1.0173,0.8810,0.8085,0.7880'
PDCalibration(InputWorkspace='calFile_8x8', TofBinning='2500,10,14500',\  PeakFunction='Gaussian',BackgroundType='Linear',\  PeakPositions=peaksToFit, CalibrationParameters='DIFC', OutputCalibrationTable='_cal',\  DiagnosticWorkspaces='_diag',MaxChiSq=7,PeakWindow=0.2,MinimumPeakHeight=20,MinimumSignalToNoiseRatio=0)
```
By changing the value of `MinimumSignalToNoiseRatio`, you can see the effect the new property has on the _cal_mask->ShowInstrument pixel map. Also, when `MinimumSignalToNoiseRatio` is > 0, the number of peaks rejected because of the low signal-to-noise ratio will be reported as a [Notice] in the log.





<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

bolotovskyr committed:
Implement signal-to-noise peak check.

bolotovskyr committed:
Improve SNR calculations. Use default SNR=0.

bolotovskyr committed:
Reinstall counts check. Reorganize pre-check reporting. Fix tests.

bolotovskyr committed:
Improve FitPeaks pre-checks. Re-enable data count check in tests.

bolotovskyr commented:
Logging the number of peaks rejected for different reasons by FitPeaks is very important for the user, so they could see the effect of the properties like `MinimumPeakHeight` or `MinimumSignalToNoiseRatio` on the data.

However, the `PDCalibration` class effectively prevents any [Notice] logging by `FitPeaks`, because of this code:

https://github.com/mantidproject/mantid/blob/c02ae4cb378648e7336cc15b6678a80eb5d20757/Framework/Algorithms/src/PDCalibration.cpp#L466

To circumvent this problem, I have used an RAII sentry object to temporarily lift the logging offset and let the [Notice]-s to be logged.

bolotovskyr committed:
Use std::accumulate

RichardWaiteSTFC commented on Framework/Algorithms/src/FitPeaks.cpp:
```diff
@@ -870,74 +930,133 @@ std::vector<std::shared_ptr<FitPeaksAlgorithm::PeakFitResult>> FitPeaks::fitPeak
+  // Zero noise usually indicates a problem with the experiment. Note that artificial data generated by
+  // FitPeaksTest might also have zero background noise.
+  double noise = estimateBackgroundNoise(vec_y);
+  if (noise <= DBL_MIN)
```
`PDCalibration` (which calls this algorithm`FitPeaks` is used at ISIS to fit peaks in a spectrum of a single detector pixel that often contains lots of zero bins  - depending on the background region chosen the standard deviation may well be zero - it looks like in this instance the noise would be set to `DBL_MAX` and therefore $signal/noise \gtrapprox 0$ - whereas in this case if the peak bins have any intensity we might want to try fitting this peak?

bolotovskyr replied:
In this case the returned DBL_MAX is actually a signal-to-noise value, not a noise value.  In other words, peaks like this will automatically pass the signal-to-noise test. Having said that, I have removed the comment referring to "zero noise" as something irregular and added a comment about the DBL_MAX.

RichardWaiteSTFC replied:
Thanks!

RichardWaiteSTFC commented on Framework/Algorithms/src/FitPeaks.cpp:
```diff
@@ -870,74 +930,133 @@ std::vector<std::shared_ptr<FitPeaksAlgorithm::PeakFitResult>> FitPeaks::fitPeak
+ */
+double estimateBackgroundNoise(const std::vector<double> &vec_y) {
+  // assume that the first and the last 10 data points are background. Combine them into a data vector.
+  assert(vec_y.size() > 20);
```
I don't think  that the fitting window is guaranteed to have at least 20 bins in it, and also it is probably not a good idea to hard-code in a number of bins because the diffractometer resolution $\Delta d/d$ is roughly constant at a given scattering angle (so atr ISIS we determine the peak window width to scale the same way) it might be better to use that first/last 10% of point or 2 bins which ever is greater?

bolotovskyr replied:
This assertion is for an earlier check of how many data points we should have in a peak window.  That check was added recently:

https://github.com/mantidproject/mantid/blob/ED_EWM_245_SNR_peak_checks/Framework/Algorithms/src/FitPeaks.cpp#L1677-L1688

One consideration is: we fit 5 parameters (3 Gaussian and 2 for linear background.)  The general rule of thumb would be to have at least 10 data points per parameter.  Another consideration is: we have an existing `estimateBackgroundParameters` method which implicitly requires to have at least 20 data points: 

https://github.com/mantidproject/mantid/blob/ED_EWM_245_SNR_peak_checks/Framework/Algorithms/src/FitPeaks.cpp#L1585-L1587

I have a question. For the type of experiments you are dealing with, would it be possible to use a smaller step in the binning, so that you have more bins per a peak window?

RichardWaiteSTFC replied:
> One consideration is: we fit 5 parameters (3 Gaussian and 2 for linear background.) The general rule of thumb would be to have at least 10 data points per parameter.

Agreed that you need more points than fitting parameters in the window (hopefully many more!) - indeed there are peak functions with more parameters (e.g. `BackToBackExponential`), so there should be a separate check for this (I thought there already was actually) which gets the number of parameters from the function chosen. If the window is too small we could skip the peak or increase the window (probably the former).

> Another consideration is: we have an existing estimateBackgroundParameters method which implicitly requires to have at least 20 data points:

It seems to me we could estimate the background parameters from fewer points and therefore throw away fewer peaks (this might be important for low d-spacing peaks which are narrow and need small windows to avoid overlap with other peak windows). 

`I have a question. For the type of experiments you are dealing with, would it be possible to use a smaller step in the binning, so that you have more bins per a peak window?`

Good question ,unfortunately this is not really an option for two reasons:

1. The minimum bin width is determined by the instrument electronics - we operate in histogram mode so data are already binned in TOF.
2. When fitting an individual detector spectrum we struggle for flux, it is then necessary to increase the bin width (typically double it)  to improve the stats for the fitting so minimizing chi-squared works better (although technically we are still be in the Poisson regime with non-normal errors) - even then we frequently have many empty bins.

RichardWaiteSTFC commented on Framework/Algorithms/src/FitPeaks.cpp:
```diff
@@ -870,74 +930,133 @@ std::vector<std::shared_ptr<FitPeaksAlgorithm::PeakFitResult>> FitPeaks::fitPeak
+  std::copy(vec_y.begin(), vec_y.begin() + 10, vec_bkg.begin());
+  std::copy(vec_y.end() - 10, vec_y.end(), back_inserter(vec_bkg));
+  // estimate the noise as the standard deviation of that combined data vector
+  auto intensityStatistics = Kernel::getStatistics(vec_bkg, StatOptions::CorrectedStdDev);
```
I'm not sure whether this standard deviation approach is valid for Poisson distributed/low counts data (regime we are often in when calling `FitPeaks` from `PDCalibration`) - perhaps it would be better to use e.g. $(1/N_{bg})\sqrt{\Sigma_{i=0}^{N_{bg}} e_i^2}$ ? Just a suggestion, it may not make much difference!

bolotovskyr replied:
I believe the formula you gave is for a standard error of the mean (SEM), i.e the mean background value.  I am using a standard deviation of the population (SD), which is the background noise. SEM is normally less than SD.

RichardWaiteSTFC replied:
Yes I suggested the SEM because i thought it might be more robust for fewer bins but agreed it is not the 'noise' on the background - which is probably the important thing for excluding peaks.

RichardWaiteSTFC replied:
A criterion which a lot of crystallographers use to discriminate valid/real peaks from spurious peaks is the `I/sigma` ratio (i.e. the integrated intensity/counts divided by the error in the integrated intensity). This would require attempting to fit the peak (for which you need to be reasonably certain there may be a peak there - e.g. from looking at signal to noise ratio pre-fitting as done here), so maybe it is a separate check post-fitting? It's just an idea and would be done separately to this PR if at all.

bolotovskyr replied:
I agree, such a post-fit check would be quite useful. And yes, if approved, it would be done separately from this PR.

bolotovskyr committed:
Better log and comments. Set SNR treshold to 0.

RichardWaiteSTFC commented:
From the docs it looks like there is already a similar parameter `FindBackgroundSigma` which is described as `Multiplier of standard deviations of the variance for convergence of peak elimination`. It is retrieved from the arguments and gets stored in `m_bkgdSimga` - but it's not clear (to me!) where that is used...

RichardWaiteSTFC commented on Framework/Algorithms/src/FitPeaks.cpp:
```diff
@@ -328,7 +375,13 @@ void FitPeaks::init() {
+                  "for pre-checking peaks before fitting, such that all the peaks with the total Y-count "
+                  "under this value will be excluded.");
+
+  declareProperty(PropertyNames::PEAK_MIN_SIGNAL_TO_NOISE_RATIO, 0.,
```
This is technically a breaking change as it will change the order of positional arguments passed in the python function  call (although I think the risk of someone calling `FitPeaks` with positional arguments as opposed to keyword arguments is low given the number of parameters). In general extra arguments should be added at the end, is there a specific reason for it to be here?

bolotovskyr replied:
Great catch! I was just trying to keep "similar things" together, but yes, I do need to move the new property to the end.

bolotovskyr replied:
As far as `FindBackgroundSigma`, it is not being used in the FitPeaks class. It's like a legacy parameter that you can set, but it doesn't affect anything.

RichardWaiteSTFC replied:
> As far as `FindBackgroundSigma`, it is not being used in the FitPeaks class. It's like a legacy parameter that you can set, but it doesn't affect anything.

Hmm, I think it would be good to remove this at some point as it could confuse the user and it doesn't do anything, but that would also be a breaking change. For now would it be possible to print a deprecation warning if this parameter is set, and suggest the user instead sets `MinimumSignalToNoiseRatio` instead?

RichardWaiteSTFC commented on Framework/Algorithms/test/FitPeaksTest.h:
```diff
@@ -943,7 +943,7 @@ class FitPeaksTest : public CxxTest::TestSuite {
     TS_ASSERT_THROWS_NOTHING(fitpeaks.setProperty("PeakFunction", "Gaussian"));
     TS_ASSERT_THROWS_NOTHING(fitpeaks.setProperty("PeakCentersWorkspace", peak_center_ws_name));
-    TS_ASSERT_THROWS_NOTHING(fitpeaks.setProperty("FitPeakWindowWorkspace", fit_window_ws_name))
+    TS_ASSERT_THROWS_NOTHING(fitpeaks.setProperty("FitPeakWindowWorkspace", fit_window_ws_name));
```
It doesn't look like there is a test that checks the algorithm for a non-default `MinimumSignalToNoiseRatio!=0` - is it possible to add a check for this where some peaks are excluded using this check?

bolotovskyr replied:
Yes, I am going to add one.

RichardWaiteSTFC reviewed (changes requested):
Sorry for the delay reviewing, just a question of the placement of the new argument in the declaration and a new test would be nice!

bolotovskyr commented:
Hi Richard, Sorry, I had to take a little break from this PR, because I was assigned to another project. But now I am back, and hoping to push some code updates very soon.  I just wanted to make sure that you are ok with the current calculation of the number of points:

bolotovskyr commented:
Hi Richard, Sorry, I had to take a little break from this PR, because I was assigned to another project. But now I am back, and hoping to push some code updates very soon. 

I just wanted to make sure we are on the same page regarding the minimum number of data points.  Are you ok with the following approach: 

- reject a peak if the peak window has fewer than **(number of fitting parameters + 2)** bins
- when it comes to the signal-to-noise calculation, require at least **23** bins, but do not reject a peak with fewer bins - just let it pass by default.  In other words, low resolution peaks in your case won't be affected.

RichardWaiteSTFC commented:
> * reject a peak if the peak window has fewer than **(number of fitting parameters + 2)** bins

Sounds good!

> * when it comes to the signal-to-noise calculation, require at least **23** bins, but do not reject a peak with fewer bins - just let it pass by default.  In other words, low resolution peaks in your case won't be affected.

OK, that works for me, but does it compromise your aims?

Also could you add a deprecation warning if `FindBackgroundSigma` is set - letting the user know it is ignored and that it will be removed at some point.

bolotovskyr commented:
> OK, that works for me, but does it compromise your aims?

It is true, this algorithm is not applicable for all cases, but I think its advantages still outweigh its limitations.  For example, we have seen its positive effect on filtering out detector areas blocked by the sample environment (PE cell).

> Also could you add a deprecation warning if FindBackgroundSigma is set - letting the user know it is ignored and that it will be removed at some point.

bolotovskyr committed:
Algorithm improvements, SNR test, deprecate FindBackgroundSigma.

bolotovskyr committed:
Merge remote-tracking branch 'origin/main' into ED_EWM_245_SNR_peak_checks

bolotovskyr committed:
Bugfix and zscore test for noise. Use 10 datapoints for noise.

bolotovskyr committed:
Address Doxygen warnings, improve comments.

bolotovskyr commented:
I have made the following major code changes:

-  changed the minimum required number of bins for fitting a peak to (number of fitting parameters + 2). 
- changed the minimum required number of bins for signal-to-noise estimation to 13. Specifically, 10 (5 leftmost + 5 rightmost) data points in the peak window will be used for the background noise estimation. This is less than 23 points I required before. I also added a z-score check for outliers to make the noise estimate more robust.
- added a deprecation warning if `FindBackgroundSigma` is set
- added a signal-to-noise test
- moved the declaration of the `MinimumSignalToNoiseRatio` property to the end of the list

I have also opened a story in our Engineering Workflow Management system: `Story 356 [Enabler] Provide a post-fitting check in PDCalibration/FitPeaks for I/Sigma ratio`.

RichardWaiteSTFC commented:
Great, at ISIS we don't have access to this so I have also made an issue for story 356 (#35224) - we're currently in a release sprint (only bug fixes) but will soon be starting new work so I may be able to pick this up if you're busy.

RichardWaiteSTFC commented on Framework/Algorithms/src/FitPeaks.cpp:
```diff
@@ -219,31 +219,46 @@ void PeakFitPreCheckResult::setNumberOfPeaksWithNotEnoughDataPoints(const size_t
+
+  // if a peak is rejected, it is rejected based on the very first check it fails
+  size_t individual_rejection_count = m_low_count_individual + m_not_enough_datapoints + m_low_snr;
+  assert(individual_rejection_count <= 1);
```
I think these `assert` statements only apply on debug builds - is that what you intended?

bolotovskyr replied:
Correct, it's a self-consistency check for debug only, to make sure the class is used as designed.

RichardWaiteSTFC commented on Framework/Algorithms/src/FitPeaks.cpp:
```diff
@@ -965,106 +983,29 @@ void rangeToIndexBounds(const vector_like &vecx, const double range_left, const
+  // 10 (a magic number) data points on both ends of the peak window to estimate the background noise
+  size_t half_number_of_bkg_datapoints{10};
+  if (vec_y.size() < 2 * half_number_of_bkg_datapoints)
+    return DBL_MIN; // this will let the peak pass the signal-to-noise check by default
```
Thanks for making this the default behaviour, I think it's safest given 10 is a 'magic-number' (though much less restrictive than the previous ~20 points required). Do you think in future perhaps we should expose this as an option to the user?

bolotovskyr replied:
Actually, the code snippet you are showing is from a previous commit.  The latest code is this: https://github.com/mantidproject/mantid/blob/ED_EWM_245_SNR_peak_checks/Framework/Algorithms/src/FitPeaks.cpp#L992-L996. 

Yes, I can add an additional property to PDCalibration and FitPeaks classes, such as `NumberOfDataPointsForPeakWindowNoiseEstimation` and set its default to 10. However, I wouldn't want the users to "break the magic" and go below 10 data points.  Also that number would have to be even.  So, all in all, I think the software would have to validate the user's input and modify it if necessary.

RichardWaiteSTFC commented on Framework/Algorithms/test/FitPeaksTest.h:
```diff
@@ -976,8 +976,104 @@ class FitPeaksTest : public CxxTest::TestSuite {
+      TS_ASSERT_EQUALS(fitted_positions_0.size(), 2); // with 2 peaks to fit
+
+      // If a check, such as signal-to-noise, fails, FitPeaks sets the peak position to -4.
+      TS_ASSERT_DELTA(fitted_positions_0[0], -4, 0);
```
Was the previous behaviour not that if a fit failed then the peak position will be its proposed peak position, while its chi-squared shall be some special value? I think changing the peak position to be negative flag (unphysical) actually makes a failure more obvious, but I wonder if it's a breaking change?

bolotovskyr replied:
No, setting the peak position to -4 has been there for 5 years: https://github.com/mantidproject/mantid/blob/main/Framework/Algorithms/src/FitPeaks.cpp#L1210-L1212.  I am just using that to see if a peak has been rejected.

RichardWaiteSTFC replied:
Fair enough, sorry!

RichardWaiteSTFC reviewed (approved):
Looks good, a useful extra option and lots of improvements to the code. Have checked and the ENGINX full and focussed calibrations work as expected. Thanks for all the changes, sorry to hold it up!

bolotovskyr commented:
No worries, all your comments were very helpful, and my code got better than the first version!

DannyHindson commented:
Please can you add a release note .rst file for this. Somewhere under `docs\source\release\v6.7.0\Diffraction`

bolotovskyr commented:
> Great, at ISIS we don't have access to your workflow management systems so I have also made an issue for story 356 (#35224) - we're currently in a release sprint (only bug fixes) but will soon be starting new work so I may be able to pick this up if you're busy.

Sure, if you want to do the work, you are welcome.  Right now this work item is in our pipeline, being reviewed, and it hasn't been approved/scheduled yet.  I will check back with you if/when it comes up for scheduling.  Or, if you can start on it earlier, please let me know as well.

mguthriem commented:
I was testing PDCalibration and noticed that it will happily fit peaks that are clearly single noise spikes. The implemented signal to noise may be under-estimating noise but, by choosing a large enough threshold for S/N , the peak will be rejected (S/N threshold of 11 or lower allowed the this peak in the figure to pass, 12 or higher rejected it). Anyway, I wondered if we might also consider post-fitting rejection on the basis of peak width. On the logic that if a peak width is < 1 bin, a peak is likely just a noise spike.

<img width="1435" alt="fit_to_noise" src="https://user-images.githubusercontent.com/90723788/233182343-132f0096-8920-4d4b-9296-c5d46f3d4bd0.png">

RichardWaiteSTFC commented:
> I was testing PDCalibration and noticed that it will happily fit peaks that are clearly single noise spikes. The implemented signal to noise may be under-estimating noise but, by choosing a large enough threshold for S/N , the peak will be rejected (S/N threshold of 11 or lower allowed the this peak in the figure to pass, 12 or higher rejected it). Anyway, I wondered if we might also consider post-fitting rejection on the basis of peak width. On the logic that if a peak width is < 1 bin, a peak is likely just a noise spike.

I have added this suggestion to the issue created about post-fit checks of I/sigma - thanks!
